File: back/src/whombat/routes/clip_predictions.py
"""REST API routes for Clip Predictions."""

from uuid import UUID

from fastapi import APIRouter, Depends

from whombat import api, schemas
from whombat.filters.clip_predictions import ClipPredictionFilter
from whombat.routes.dependencies import Session
from whombat.routes.types import Limit, Offset

__all__ = [
    "clip_predictions_router",
]


clip_predictions_router = APIRouter()


@clip_predictions_router.post(
    "/",
    response_model=schemas.ClipPrediction,
)
async def create_clip_prediction(
    session: Session,
    clip_uuid: UUID,
    data: schemas.ClipPredictionCreate,
):
    """Create a clip prediction."""
    clip = await api.clips.get(session, clip_uuid)
    clip_prediction = await api.clip_predictions.create(
        session,
        clip=clip,
    )
    for predicted_tag in data.tags:
        tag = await api.tags.get_or_create(
            session,
            key=predicted_tag.tag.key,
            value=predicted_tag.tag.value,
        )
        clip_prediction = await api.clip_predictions.add_tag(
            session,
            clip_prediction,
            tag,
            predicted_tag.score,
        )
    await session.commit()
    return clip_prediction


@clip_predictions_router.get(
    "/",
    response_model=schemas.Page[schemas.ClipPrediction],
)
async def get_clip_predictions(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    filter: ClipPredictionFilter = Depends(ClipPredictionFilter),  # type: ignore
    sort_by: str = "-created_on",
):
    """Get a page of clip predictions."""
    (
        clip_predictions,
        total,
    ) = await api.clip_predictions.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=clip_predictions,
        total=total,
        limit=limit,
        offset=offset,
    )


@clip_predictions_router.delete(
    "/detail/",
    response_model=schemas.ClipPrediction,
)
async def delete_clip_prediction(
    session: Session,
    clip_prediction_uuid: UUID,
):
    """Delete a clip prediction."""
    clip_prediction = await api.clip_predictions.get(
        session,
        clip_prediction_uuid,
    )
    clip_annotaiton = await api.clip_predictions.delete(
        session,
        clip_prediction,
    )
    await session.commit()
    return clip_annotaiton


@clip_predictions_router.post(
    "/detail/tags/",
    response_model=schemas.ClipPrediction,
)
async def add_predicted_tag(
    session: Session,
    clip_prediction_uuid: UUID,
    key: str,
    value: str,
    score: float,
):
    """Add a tag to a clip prediction."""
    clip_prediction = await api.clip_predictions.get(
        session,
        clip_prediction_uuid,
    )
    tag = await api.tags.get(session, (key, value))
    clip_prediction = await api.clip_predictions.add_tag(
        session,
        clip_prediction,
        tag,
        score,
    )
    await session.commit()
    return clip_prediction


@clip_predictions_router.delete(
    "/detail/tags/",
    response_model=schemas.ClipPrediction,
)
async def remove_predicted_tag(
    session: Session,
    clip_prediction_uuid: UUID,
    key: str,
    value: str,
):
    """Remove a tag from a clip prediction."""
    clip_prediction = await api.clip_predictions.get(
        session,
        clip_prediction_uuid,
    )
    tag = await api.tags.get(session, (key, value))
    clip_prediction = await api.clip_predictions.remove_tag(
        session,
        clip_prediction,
        tag,
    )
    await session.commit()
    return clip_prediction

File: front/src/lib/api/clip_predictions.ts
import { AxiosInstance } from "axios";

import { GetMany, Page } from "@/lib/api/common";
import * as schemas from "@/lib/schemas";
import type * as types from "@/lib/types";

const DEFAULT_ENDPOINTS = {
  create: "/api/v1/clip_predictions/",
  getMany: "/api/v1/clip_predictions/",
  delete: "/api/v1/clip_predictions/detail/",
  addTag: "/api/v1/clip_predictions/detail/",
  removeTag: "/api/v1/clip_predictions/detail/",
};

export function registerClipPredictionsAPI(
  instance: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function create(
    clipPrediction: types.ClipPrediction,
    data: types.ClipPredictionCreate,
  ): Promise<types.ClipPrediction> {
    const body = schemas.ClipPredictionCreateSchema.parse(data);
    const response = await instance.post(endpoints.create, body, {
      params: { clip_prediction_uuid: clipPrediction.uuid },
    });
    return schemas.ClipPredictionSchema.parse(response.data);
  }

  async function getMany(
    query: types.GetMany & types.ClipPredictionFilter,
  ): Promise<types.Page<types.ClipPrediction>> {
    const params = GetMany(schemas.ClipPredictionFilterSchema).parse(query);
    const response = await instance.get(endpoints.getMany, {
      params: {
        limit: params.limit,
        offset: params.offset,
        sort_by: params.sort_by,
        clip__eq: params.clip?.uuid,
        recording__eq: params.recording?.uuid,
        tag__key: params.tag?.tag.key,
        tag__value: params.tag?.tag.value,
        tag__gt: params.tag?.gt,
        tag__lt: params.tag?.lt,
        sound_event_tag__key: params.sound_event_tag?.tag.key,
        sound_event_tag__value: params.sound_event_tag?.tag.value,
        sound_event_tag__gt: params.sound_event_tag?.gt,
        sound_event_tag__lt: params.sound_event_tag?.lt,
        model_run__eq: params.model_run?.uuid,
        user_run__eq: params.user_run?.uuid,
      },
    });
    return Page(schemas.ClipPredictionSchema).parse(response.data);
  }

  async function deleteClipPrediction(
    clipPrediction: types.ClipPrediction,
  ): Promise<types.ClipPrediction> {
    const { data } = await instance.delete(endpoints.delete, {
      params: { clip_prediction_uuid: clipPrediction.uuid },
    });
    return schemas.ClipPredictionSchema.parse(data);
  }

  async function addTag(
    clipPrediction: types.ClipPrediction,
    tag: types.Tag,
    score: number,
  ): Promise<types.ClipPrediction> {
    const { data } = await instance.post(
      endpoints.addTag,
      {},
      {
        params: {
          clip_prediction_uuid: clipPrediction.uuid,
          key: tag.key,
          value: tag.value,
          score: score,
        },
      },
    );
    return schemas.ClipPredictionSchema.parse(data);
  }

  async function removeTag(
    clipPrediction: types.ClipPrediction,
    tag: types.Tag,
  ): Promise<types.ClipPrediction> {
    const { data } = await instance.delete(endpoints.removeTag, {
      params: {
        clip_prediction_uuid: clipPrediction.uuid,
        key: tag.key,
        value: tag.value,
      },
    });
    return schemas.ClipPredictionSchema.parse(data);
  }

  return {
    create,
    getMany,
    delete: deleteClipPrediction,
    addTag,
    removeTag,
  } as const;
}
